File: test/adventofcode/year2021/day05_test.clj
(ns adventofcode.year2021.day05-test
  (:require [clojure.test :refer :all]
            [adventofcode.year2021.day05 :as day]))

(def example-input
  (day/parse-input
    "0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2"))

(deftest part1-example
  (is (= (day/part1 example-input) 5)))

(deftest part2-example
  (is (= (day/part2 example-input) :todo)))

File: 2025/typescript/src/day05.test.ts
import {describe, expect, test} from "vitest";
import {input} from "./input";

const EXAMPLE = `3-5\n10-14\n16-20\n12-18\n\n1\n5\n8\n11\n17\n32`;

const part1 = (input: string) => {
  const parts = input.split("\n\n");
  const ranges = parseRanges(parts[0]);
  const ingredients = parts[1].split("\n").map(line => parseInt(line));

  return ingredients.filter(i => ranges.find(([first, last]) => i >= first && i <= last) != undefined).length;
};

const part2 = (input: string) => {
  const ranges = parseRanges((input.split("\n\n"))[0]).toSorted(([f1,], [f2,]) => f1 - f2);

  return ranges.reduce(
    ([count, prev_last], [first, last]) => {
      if (first > prev_last) return [count + (last - first + 1), last];
      if (last > prev_last) return [count + (last - prev_last), last];
      return [count, prev_last];
    }, [0, -1]
  )[0];
};

const parseRanges = (input: string): [number, number][] =>
  input.split("\n").map(line => {
    const fl = line.split("-");
    return [parseInt(fl[0]), parseInt(fl[1])]
  });

describe('Day 05', () => {
  test('Part 1', async () => {
    expect(part1(EXAMPLE)).toEqual(3);
    expect(part1(await input(5))).toEqual(601);
  });

  test('Part 2', async () => {
    expect(part2(EXAMPLE)).toEqual(14);
    expect(part2(await input(5))).toEqual(367899984917516);
  });
});
